### PR TITLE
Introduce integration test suite for OpenShift example API

### PR DESCRIPTION
**Pull request by JoelSpeed**
Marked as WIP as I want to add an integration prow job separately from the unit prow job

We commonly have scenarios in API reviews where we don't have the ability to validate the serialization and end to end behaviour of the validations we are suggesting.

In an attempt to make that easier, this PR sets up an integration test suite using envtest (runs a Kube API and etcd as part of the test suite) which validates for the example API each of the API examples we are trying to demonstrate with that API.

Additionally, it adds an example of a discriminated union validated with CEL as this is needed for some of our open API reviews, and tests that prove the validation is working as we expect.

I hope over time to work with API authors to add additional tests for their APIs to prove the validations we are adding are working as expected, in an effort to reduce the manual validation that we currently ask for on API PRs.

Note: The tests run as a different module so the new deps do not affect those importing our APIs.

I should probably also add myself as owner for this module 🤔

### CODE REVIEW

**Comment by JoelSpeed** [comment]
JUnit report here shows the tests ran and passed, https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_api/1298/pull-ci-openshift-api-master-unit/1570384423361187840

File: Makefile [hunk 1]
```diff
@@ -111,3 +111,11 @@ verify-with-container:
...
 generate-with-container:
 	$(RUNTIME) run -ti --rm -v $(PWD):/go/src/github.com/openshift/api:z -w /go/src/github.com/openshift/api $(RUNTIME_IMAGE_NAME) make update
+
+
+#DO NOT MERGE
+test-unit: integration
+
+.PHONY: integration
+integration:
+	make -C tests integration
```
Comment (deads2k) [inline Makefile:121]: what does `make -C` do?

Comment (JoelSpeed) [reply Makefile:121]: Changes folder before executing the target, so this is, change to the tests dir and then execute it's integration target

**Comment by JoelSpeed** [comment]
Created a PR to the release repo to add the integration test check, this will need to merge first though else we will block PRs
https://github.com/openshift/release/pull/32327

File: example/v1/types_stable.go [hunk 2]
```diff
@@ -63,6 +67,40 @@ const (
 	TechPreviewOnlyValue EvolvingDiscriminator = "TechPreviewOnlyValue"
 )
 
+// CELUnion demonstrates how to use a discriminated union and how to validate it using CEL.
```
Comment (deads2k) [inline example/v1/types_stable.go:70]: if the RequiredMember is set, you want to verify the self.type.

Comment (JoelSpeed) [reply example/v1/types_stable.go:70]: First part of the ternary condition, `has(self.type)` fails if type isn't set, at which point we fall into `!has(self.requiredMember)` which fails if `RequiredMember` is set, which then fails the validation and returns the error.

Comment (JoelSpeed) [reply example/v1/types_stable.go:70]: We have `should not allow a member without the optional type value` as a test that checks the optional but I can add for required too

**Comment by JoelSpeed** [comment]
Based on earlier test runs I'm happy that this is working. The last push dropped the WIP content so I've marked this as ready.

Once this is merged I'll seek to get the release repo PR merged which will enable this check across the repo.

Last push included explicit tests for both optional and required members being set without the type, being an error

File: example/v1/types_stable.go (hunk 2 above)
Comment (JoelSpeed) [reply example/v1/types_stable.go:70]: Added explicit tests for both now

File: tests/example/stable.go [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+
+	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/utils/pointer"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/envtest"
+	"sigs.k8s.io/controller-runtime/pkg/envtest/komega"
+)
+
+var _ = Describe("[Stable] Example API", func() {
```
Comment (deads2k) [inline tests/example/stable.go:19]: These all appear to follow a pattern of either
1. create
2. expected error

or
1. create 
2. expect success

or
1. create
2. expect success
3. update
4. failure message


Given the suggested fanout from these integration tests, I think your time would be well spent to allow directories to be created with content with files like create,  expected_create_failure, update_1, expected_update_1_failure and auto-run the tests based on those.  That will help you get traction for your common cases.

File: machine/v1/control-plane-machine-set.aws.stable.testsuite.yaml [hunk 4]
```diff
@@ -0,0 +1,73 @@
+apiVersion: apiextensions.k8s.io/v1 # Hack because controller-gen complains if we don't have this
```
Comment (deads2k) [inline machine/v1/control-plane-machine-set.aws.stable.testsuite.yaml:1]: Grab someone on the machine API team to review please and ask them to imagine dozens of validation tests in this file.

**Comment by deads2k** [comment]
Minor on test name.  Can you find a way to add group, version, and resource to the test name so everything sorts well.

I don't personally care for the single file approach, but I'm willing to start there if you like it.  The testing result here is great.

Before merge this we need a PR that demonstrates some tests failing.

largely lgtm.

File: machine/v1/control-plane-machine-set.aws.stable.testsuite.yaml (hunk 4 above)
Comment (JoelSpeed) [reply machine/v1/control-plane-machine-set.aws.stable.testsuite.yaml]: I've split the MAPI parts of this into #1297 which I'll merge separately once this is in. Got a few more bits I want to explore there

**Comment by JoelSpeed** [comment]
> Minor on test name. Can you find a way to add group, version, and resource to the test name so everything sorts well.

I've added `[groupVersionResource.String()]` as a prefix to each test so we should be able to sort nicely now

> Before merge this we need a PR that demonstrates some tests failing.

I'll push some failures onto the CPMS PR

I've also added the ability to have `onUpdate` tests so we can prove that validations of things like immutable fields work

**Comment by JoelSpeed** [comment]
I introduced some [deliberate failures](https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_api/1297/pull-ci-openshift-api-master-unit/1572591972550447104) on the other PR.

Some interesting examples:
* When there's no error but we expected an error, the current message from JUnit looks like: `Expected an error, got nil failed /go/src/github.com/openshift/api/tests/generator.go:167`
* When there's a mismatch between the error types, things are a bit more complicated, but you can see the messages don't line up:
  ```
  Expected
    <*errors.StatusError | 0xc0002f8dc0>: {
        ErrStatus: {
            TypeMeta: {Kind: "Status", APIVersion: "v1"},
            ListMeta: {
                SelfLink: "",
                ResourceVersion: "",
                Continue: "",
                RemainingItemCount: nil,
            },
            Status: "Failure",
            Message: "ControlPlaneMachineSet.machine.openshift.io \"test-kjnkv\" is invalid: spec.template.machines_v1beta1_machine_openshift_io.metadata.labels: Invalid value: \"object\": label 'machine.openshift.io/cluster-api-machine-role' is required, and must have value 'master'",
            Reason: "Invalid",
            Details: {
                Name: "test-kjnkv",
                Group: "machine.openshift.io",
                Kind: "ControlPlaneMachineSet",
                UID: "",
                Causes: [
                    {
                        Type: "FieldValueInvalid",
                        Message: "Invalid value: \"object\": label 'machine.openshift.io/cluster-api-machine-role' is required, and must have value 'master'",
                        Field: "spec.template.machines_v1beta1_machine_openshift_io.metadata.labels",
                    },
                ],
                RetryAfterSeconds: 0,
            },
            Code: 422,
        },
    }
  to match error
    <*matchers.ContainSubstringMatcher | 0xc000509da0>: {
        Substr: "spec.template.machines_v1beta1_machine_openshift_io.metadata.labels: Invalid value: \"object\": label 'machine.openshift.io/cluster-api-machine-role' is required, and must have value 'master' <broken>"
  ```
* When there's a mismatch between the expected fields and the fields that persisted, it will tell you which fields differed, and then give you the full objects:
  ```
  the following fields were expected to match but did not:
  [(spec.replicas/spec.replicas) (spec.strategy/spec.strategy)]
   ```

**Comment by deads2k** [comment]
I don't see the tests added for the example API group here in the unit test results in https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_api/1298/pull-ci-openshift-api-master-unit/1572532119568125952

**Comment by deads2k** [comment]
I'm ok with the changes, but I'm not seeing these tests run, so something appears to be missing.

**Comment by JoelSpeed** [comment]
I had been running the tests under the `unit` target, but I think having a separate status check would be better.
So I dropped the temporary make adjustments from this PR and was intending to merge https://github.com/openshift/release/pull/32327 once this goes in.

The temporary make adjustment is still present on #1297 so you can see the tests are running there

**Pull request merged**